Enhance filename validation

cocopaw opened the pull request:
# Enhance filename validation

This PR extends [#23060] and implements the following **key improvements**:

1. **Refactor `isValidFileName`**
2. **Enhance `toValidFileName`** to use new filename validation logic
3. **Enhance `Path::isValid`** to use new filename validation logic
4. **Block colon (`:`) on macOS****

---

**The updated validation and sanitization logic is now applied consistently throughout, including:**

- File and folder renaming in the GUI (Content pane) and WebUI/API
- Torrent creation (destination path validation and temporary file handling)
- Torrent addition (Save path validity checking)
- Category management (save paths and incomplete torrents save paths)
- User profile creation
- RSS feed legacy filename generation
- Torrent file exporting

---

### Testing
**Tested:** Windows 11 + Linux Mint 22.1
**Note:** Unable to test the macOS-specific colon (`:`) blocking due to lack of access.

---

** **The colon char (`:`) is reserved in macOS Finder/GUI as it is a legacy HFS path separator.**

**References:**
- [Inside Macintosh: Files - File Manager (PDF, p. 2-27)](https://developer.apple.com/library/archive/documentation/mac/pdf/Files/File_Manager.pdf)  
  ("_File and directory names consist of any sequence of 1 to 31 characters, excluding colons_")
- [Apple Support: Rename files, folders, and disks on Mac](https://support.apple.com/guide/mac-help/rename-files-folders-and-disks-on-mac-mchlp1144/mac)  
  ("_You can't include a colon (:) or start the name with a period (.)_")
- [Apple Filing Protocol Concepts](https://developer.apple.com/library/archive/documentation/Networking/Conceptual/AFP/Concepts/Concepts.html)  
  ("_AFP Long Names are made up of at most 31 characters; valid characters are any printable ASCII code except colon (0x3A) and null (0x00)._")

---

Fixes #22426, Fixes #21899, Fixes #16906, Fixes #16834, Fixes #16271, Fixes #11276, Fixes #9085, Fixes #11340, Fixes #14728, Fixes #20337, Fixes #17049, Fixes #17469, Fixes #15378, Fixes #15227, Fixes #14908, Fixes #20742, Fixes #20412

cocopaw committed:
Refactor isValidName. Enahnce toValidFileName

cocopaw committed:
Improve torrent file path sanitization in TorrentCreator

cocopaw committed:
Enhance toValidPath to match toValidFileName sanitization

cocopaw committed:
Refactor to reduce code duplication. Rename isValidName to isValidFileName

cocopaw committed:
Enhance "." check and remove toValidPath

cocopaw committed:
Minor formatting changes

cocopaw committed:
Fix CON_1CON bug

glassez commented on src/base/bittorrent/sessionimpl.cpp:
```diff
@@ -970,7 +970,7 @@ Path SessionImpl::categoryDownloadPath(const QString &categoryName, const Catego
     const Path path = !downloadPathOption.path.isEmpty()
             ? downloadPathOption.path
-            : Utils::Fs::toValidPath(name); // use implicit download path
+            : Path(Utils::Fs::toValidFileName(name)); // use implicit download path
```
Here too.

glassez commented on src/base/bittorrent/sessionimpl.cpp:
```diff
@@ -940,12 +940,12 @@ Path SessionImpl::categorySavePath(const QString &categoryName, const CategoryOp
         else
         {
-            path = Utils::Fs::toValidPath(categoryName);
+            path = Path(Utils::Fs::toValidFileName(categoryName));
```
> I couldnt fault itbwhen I was testing. I also tested subcategories. Can u give me an example scenario that would break it?

Did you test it with "flat" categories (i.e. when subcategories is disabled)?

cocopaw replied:
I tested it with both flat categories and subcategories.

glassez replied:
> 1. Removed ":" from the char set as this would cause issues if the path had a win root like "C:\".

What if ":" appears in other place of input string?
I would postpone any changes to this feature until you have completely reworked it.

cocopaw replied:
Ok give me a few days and ill see what I can do.

Are you ok with Path.isvalid() being moved to fs utils and reworked too ?

glassez replied:
I don't mind.

cocopaw committed:
Remove *Internal from isValidFileName/toValidFileName

cocopaw committed:
Reinstate toValidPath

cocopaw committed:
Update src/base/utils/fs.cpp

Co-authored-by: Vladimir Golovnev <glassez@yandex.ru>

cocopaw committed:
Move Path::isValid() to Utils::Fs::isValidPath

Chocobo1 commented on src/base/utils/fs.cpp:
```diff
@@ -62,6 +62,35 @@
+        if ((unicode < 32) || (unicode == 127) || (c == u'/'))
+            return true;
+#ifdef Q_OS_WIN
+        static const QSet<QChar> reservedChars{u'\\', u'<', u'>', u':', u'"', u'|', u'?', u'*'};
```
I would use a `switch` for even lower perf overhead:
```suggestion
        switch (c)
        {
        case u'\\':
        case u'<':
        // ...
            return true;
        };
        return false;
```

Chocobo1 replied:
I'm aware that `switch` might take too much vertical space and one would prefer it to be concise. So, I don't strongly insist about it.
On the other hand, I really dislike wasting CPU cycles. A little bit of waste here and there then overall slows down. I would still suggest keeping the previous `if` comparison. IMO it is worth trading a little bit of readability for better performance code.

cocopaw replied:
Isn't this just used infrequently for renaming though? If it were a core hot-loop I'd def be keen on the swtich version for performance too.

I asked xAi to analyse the performace gains at it came back with this:

`Key takeaway: The switch is ~0.05–0.15 μs faster per check—utterly negligible for renaming (e.g., <0.02 ms even for a 100-char filename). Readability wins with QSet.`

Given that this function is infrequent and therefor not performance critical my prefence is still for the neater and easier to read and maintain QSet. 

That being said, if you still "_**really dislike**_" the idea (i.e. if you have somewhat strong feeling on it), then I'm happy to go with what you want. 

The final decision is yours. Please let me know what you would like to do here. 

For reference, the switch version that I have looks like this:

```
// Shared check if a character is reserved (Control, DEL, '/', or Windows-specific)
    bool isReservedCharacter(const QChar c)
    {
        const ushort unicode = c.unicode();
        if ((unicode < 32) || (unicode == 127) || (c == u'/'))
            return true;
#ifdef Q_OS_WIN
   switch (c.unicode()) {
        case u'\\':
        case u'<':
        case u'>':
        case u':':
        case u'"':
        case u'|':
        case u'?':
        case u'*':
            return true;
    }
        return false;
#elif defined(Q_OS_MACOS)
        return (c == u':');
#else
        return false;
#endif
    }
```


vs the existing QSet version:


   ```
 // Shared set of reserved device names for Windows
    const QSet<QString> reservedDeviceNames
    {
        u"CON"_s, u"PRN"_s, u"AUX"_s, u"NUL"_s,
        u"COM1"_s, u"COM2"_s, u"COM3"_s, u"COM4"_s,
        u"COM5"_s, u"COM6"_s, u"COM7"_s, u"COM8"_s,
        u"COM9"_s, u"COM¹"_s, u"COM²"_s, u"COM³"_s,
        u"LPT1"_s, u"LPT2"_s, u"LPT3"_s, u"LPT4"_s,
        u"LPT5"_s, u"LPT6"_s, u"LPT7"_s, u"LPT8"_s,
        u"LPT9"_s, u"LPT¹"_s, u"LPT²"_s, u"LPT³"_s
    };
#endif

    // Shared check if a character is reserved (Control, DEL, '/', or Windows-specific)
    bool isReservedCharacter(const QChar c)
    {
        const ushort unicode = c.unicode();
        if ((unicode < 32) || (unicode == 127) || (c == u'/'))
            return true;
#ifdef Q_OS_WIN
        const QSet<QChar> reservedChars {u'\\', u'<', u'>', u':', u'"', u'|', u'?', u'*'};
        return reservedChars.contains(c);
#elif defined(Q_OS_MACOS)
        return (c == u':');
#else
        return false;
#endif
    }
```

cocopaw committed:
Rewrite Utils::Fs::isValidPath

cocopaw commented on src/base/utils/fs.cpp:
```diff
@@ -62,6 +62,35 @@
+        if ((unicode < 32) || (unicode == 127) || (c == u'/'))
+            return true;
+#ifdef Q_OS_WIN
+        static const QSet<QChar> reservedChars{u'\\', u'<', u'>', u':', u'"', u'|', u'?', u'*'};
```
Done

cocopaw commented on src/base/utils/fs.cpp:
```diff
@@ -186,19 +215,92 @@ bool Utils::Fs::sameFiles(const Path &path1, const Path &path2)
+                inReservedSequence = false;
+            }
+    });
+    validName = std::move(newName);
```
Done

cocopaw committed:
Chocobo1 review requests

cocopaw committed:
Rewrite Utils::Fs::toValidPath

cocopaw committed:
Remove Utils::Fs::isValidFileName from TorrentsController::renameFileAction

cocopaw committed:
Silence unused parameter warning in isDriveLetterPath (non-Windows platforms)

cocopaw committed:
Fix Path::isValid() failing on "/"

cocopaw committed:
Add ":" checking in OSX

cocopaw committed:
Use 'Q_OS_MACOS' instead of 'Q_OS_OSX' (depricated)

Chocobo1 commented on src/base/bittorrent/torrentcreator.cpp:
```diff
@@ -232,16 +233,23 @@ void TorrentCreator::run()
 
         checkInterruptionRequested();
 
+        // Save the generated torrent data to the specified path, sanitizing the filename and falling back to a temporary file if invalid
```
```suggestion
```

cocopaw commented on src/webui/api/torrentscontroller.cpp:
```diff
@@ -1990,10 +1990,6 @@ void TorrentsController::renameFileAction()
-    const QString newFileName = QFileInfo(params()[u"newPath"_s]).fileName();
-    if (!Utils::Fs::isValidName(newFileName))
-        throw APIError(APIErrorType::Conflict, tr("File name has invalid characters"));
-
```
Becuase it's a redundant check now. It is already checked downstream via `BitTorrent::AbstractFileStorage::renameFile`. Now that Path.isValid() is using the new logic we not longer need to check it via `Utils::Fs::isValidName `in `torrentscontroller.cpp`. 

The error checking is also better via `BitTorrent::AbstractFileStorage::renameFile `as it is more descriptive (it specifies if the oldpath or newpath is invalid instead of just the whole thing)

Process Flow:

`TorrentsController::renameFileAction() -> torrent->renameFile -> BitTorrent::AbstractFileStorage::renameFile -> oldPath.isValid()`

```
void BitTorrent::AbstractFileStorage::renameFile(const Path &oldPath, const Path &newPath)
{
    if (!oldPath.isValid())
        throw RuntimeError(tr("The old path is invalid: '%1'.").arg(oldPath.toString()));
    if (!newPath.isValid())
        throw RuntimeError(tr("The new path is invalid: '%1'.").arg(newPath.toString()));
    if (newPath.isAbsolute())
        throw RuntimeError(tr("Absolute path isn't allowed: '%1'.").arg(newPath.toString()));
```

My previous comment (for reference):

```
For the commit "Remove Utils::Fs::isValidFileName from TorrentsController::renameFileAction", 
the functions below invoke "torrent->renameFile", which now performs validation via the 
new "Path.isValid()". The prior "isValidName" checks in torrentcontentmodel.cpp are therefore 
redundant and have been removed.

src/gui/torrentcontentmodel.cpp
..
TorrentsController::renameFileAction
TorrentsController::renameFolderAction

```

cocopaw committed:
Chocobo1 review changes

cocopaw committed:
Fix CI error "isReservedCharacter not all control paths return a value"

cocopaw commented:
@Chocobo1 @glassez , is this still under review?

cocopaw committed:
Merge branch 'master' into enhance_filename_validation

glassez commented:
@cocopaw 
At least I would expect to have the second overload taking path parameter of type Path, e.g.:
```c++
bool Utils::Fs::isValidPath(const Path &path)
{
    return isValidPath(path.data());
}
```

Chocobo1 commented on src/base/path.cpp:
```diff
@@ -85,32 +78,7 @@ Path::Path(const std::string &pathStr)
-    const QRegularExpression regex {u"\\0"_s};
-#endif
-    return !m_pathStr.contains(regex);
+    return Utils::Fs::isValidPath(*this);
```
>I could just remove path.isValid() entirely from path.h and path.cpp and then change whatever was calling that to call Utils::Fs::isValidPath instead if preffered?

Yes that should be done.

cocopaw commented:
After further consideration, I think we should:

- Keep `Utils::Fs::isValidFileName` and `Utils::Fs::toValidFileName` as **QString-only** (they're inherently single-component / string-level operations)
- Make `Utils::Fs::isValidPath` and `Utils::Fs::toValidPath` **Path-only** (single signature, no QString version)

This mirrors the pattern already used for most other `Utils::Fs` functions that operate on full paths (`isDir`, `mkpath`, `renameFile`, `freeDiskSpaceOnPath`, etc.), while keeping the pure-string helpers (`isValidFileName`, `toValidFileName`) separate.

This should keep things simpler, cleaner and avoids any unessesary complication with overloading. Do you agree?

cocopaw committed:
Replace Path::isValid() with Utils::Fs::isValidPath()

cocopaw committed:
Revert "Remove Utils::Fs::isValidFileName from TorrentsController::renameFileAction"

This reverts commit bdd88dea131a11d15e3cd56e884c3b7a859065f6.

glassez commented:
> Let me know your thoughts

In general, I agree.
>if you agree, I'll go ahead and reimplement it like this.

Let's wait a little longer to see what @Chocobo1 says.

P.S.
How about https://github.com/qbittorrent/qBittorrent/pull/23386#issuecomment-4074173006?

cocopaw committed:
Update src/base/utils/fs.cpp

Co-authored-by: Vladimir Golovnev <glassez@yandex.ru>

cocopaw committed:
Update src/base/utils/fs.cpp

Co-authored-by: Vladimir Golovnev <glassez@yandex.ru>

cocopaw committed:
Use Qt::SkipEmptyParts in Utils::Fs::toValidPath

cocopaw committed:
Use QStringView in Utils::Fs::toValidFileName

cocopaw commented on src/base/utils/fs.cpp:
```diff
@@ -186,90 +221,174 @@ bool Utils::Fs::sameFiles(const Path &path1, const Path &path2)
+    QString newName;
+    newName.reserve(validName.size());
+    bool inReservedSequence = false;
+    for (const QChar c : asConst(validName))
```
Done.

cocopaw committed:
Update comments

cocopaw committed:
Use QStingView where possible in Fs

cocopaw committed:
Relocate isValidPath+toValidPath to Path class

cocopaw committed:
Refine isValid + makeValidPath in Path

cocopaw committed:
Refine isReservedCharacter

cocopaw committed:
Fix TestPath  failing on "/" and ""

cocopaw committed:
Make TorrentCreator use Path::makeValidPath

cocopaw committed:
Remove fs.h dependency

cocopaw committed:
Revise comments in Utils::Fs::isValidFileName

cocopaw committed:
Convert makeValidPath to QStringView

cocopaw committed:
Refine isValidFileName call in TorrentsController

cocopaw committed:
Minor formatting changes to makeValidPath

cocopaw committed:
Merge branch 'qbittorrent:master' into enhance_filename_validation

cocopaw committed:
Refine comments and structure in Path::isValid

cocopaw committed:
Add early file validation to TorrentCreator

cocopaw committed:
Refine isReservedDeviceName definitiion

Co-authored-by: Vladimir Golovnev <glassez@yandex.ru>

cocopaw commented on src/base/utils/fs.cpp:
```diff
@@ -62,6 +62,57 @@
+        const ushort unicode = c.unicode();
+        if ((unicode < 32) || (unicode == 127) || (c == u'/'))
+            return true;
+#ifdef Q_OS_WIN
```
Wouldn't it be better to change the four #if defined(Q_OS_XXX) to #ifdef Q_OS_XXX instead?

The [Coding Guidelines](https://github.com/qbittorrent/qBittorrent/blob/master/CODING_GUIDELINES.md) seem to recommend #ifdef for OS checks, as shown in the example below:

```
// System headers
#ifdef Q_OS_WIN
#include <windows.h>
#endif
```

glassez replied:
> The [Coding Guidelines](https://github.com/qbittorrent/qBittorrent/blob/master/CODING_GUIDELINES.md) seem to recommend #ifdef for OS checks, as shown in the example below:
> 
> ```
> // System headers
> #ifdef Q_OS_WIN
> #include <windows.h>
> #endif
> ```

Yes.
But it is unlikely about the case when it is followed by `#elif defined(Q_OS_XXX)` which looks ugly/inconsistent, IMO:
 ```
// System headers
#ifdef Q_OS_WIN
#include <windows.h>
#elif defined(Q_OS_MAC)
#include <some_header.h>
#endif
> ```

cocopaw committed:
Remove additional QStringView from toValidFileName

cocopaw committed:
Add whitespace check to the end of toValidFileName

cocopaw commented on src/base/utils/fs.cpp:
```diff
@@ -239,37 +260,72 @@ bool Utils::Fs::isValidName(const QString &name)
+            validName += c;
+            inReservedSequence = false;
+        }
+    };
```
Good point. I've added trimmed at the end before validName is retured.

cocopaw committed:
Reorder QStingView creation in isValid()

Co-authored-by: Vladimir Golovnev <glassez@yandex.ru>

cocopaw committed:
Update consistency of #ifdef in isReservedCharacter

cocopaw committed:
Remove duplicate hasDriveLetter call

cocopaw committed:
Fix "#ifdef defined" bug

cocopaw commented on src/base/path.cpp:
```diff
@@ -302,6 +298,39 @@ Path Path::commonPath(const Path &left, const Path &right)
+    QStringList validComponents;
+    validComponents.reserve(components.size());
+    for (const QStringView &comp : components)
+        validComponents << Utils::Fs::toValidFileName(comp, pad);
```
I'm strongly against this idea. I believe coupling the logic of both functions into makeValidPath is unnecessary and would make it much harder to evaluate behavior and correctness. Adding an isValid check would also increase compute and complexity. The current code is simple, keeps the responsibilities separate, and is already specifically tested and working correctly for its intended use (category save path checking).

cocopaw replied:
Ok noted. I get your point now.  

I'm going to be re-working this secion as a private helper anyway (as discussed), so lets see where we end up after that.

cocopaw committed:
Make makeValidPath use const QStringView

cocopaw commented on src/base/bittorrent/torrentcreator.cpp:
```diff
@@ -249,14 +249,16 @@ void TorrentCreator::run()
         const auto result = std::invoke([torrentFilePath = m_params.torrentFilePath, entry]() -> nonstd::expected<Path, QString>
         {
-            if (!torrentFilePath.isValid())
+            const Path finalTorrentFilePath = Path::makeValidPath(torrentFilePath.data());
```
Ok noted. I'll re-work it as a private function and check what you mention. 

FYI, you may not have had a change to read the revised first part of my original comment as I wrote as I edited it with additional info after you had already replied.

cocopaw replied:
It turns out we don't really need `makeValidPath/toValidPath` at all. `categorySavePath `in `sessionimpl `only was only ever recursively requesting sanitisation of in individual category name (i.e. not the entire path). So we can just use `toValidFileName` instead. This makes things a lot simpler. I've implemented and tested it in this way now. 

This seems to work well as even if the user enters a complete mess of windows reserved device names and one or more reserved chars, it just allows it in the GUI but sanitizes the save path. 

For example here the user has entered the Win reserved device name `con` and also many win reserved chars `<<<` :


<img width="490" height="424" alt="image" src="https://github.com/user-attachments/assets/0842628e-9dfd-43c5-90e0-109970193c71" />

-> 

<img width="346" height="170" alt="image" src="https://github.com/user-attachments/assets/4d7ec216-95ad-4ee6-b3f5-6ab871bbd564" />

cocopaw committed:
Add multiple separator tests to Path

cocopaw committed:
Remove makeValidPath

cocopaw committed:
Allow . and .. in path components

cocopaw commented on test/testpath.cpp:
```diff
@@ -74,6 +74,9 @@ private slots:
         QCOMPARE(Path(uR"(\)"_s).isValid(), true);
 
+        QCOMPARE(Path(u"."_s).isValid(), false);
+        QCOMPARE(Path(u".."_s).isValid(), false);
```
I've updated `Path::isValid()` so `.` and `..` are now allowed as path components (including in relative paths like `../foo` or `a/../b`).

This required adding of `Utils::Fs::isValidPathComponent()` as a slightly more lenient variant which allows "." and ".." used for full path validation.

`Utils::Fs::isValidFileName()` remains strict (rejects `.` and `..`) because it’s used for actual filenames (GUI rename, etc.). 

`TestPath::testIsValid()` has been expanded with comprehensive relative path cases. All tests also pass against current master.

cocopaw committed:
Change mid to sliced in Path::isValid

cocopaw committed:
Optimize Reserved Char pad in toValidFileName

cocopaw committed:
Minor comment and formatting changes

cocopaw committed:
Relocate Path::isValid()

cocopaw committed:
Minor formatting changes to Utils::Fs

cocopaw commented:
@glassez This is ready for review.

Please also see my responses to your previous comments in the conversations above.

glassez commented on src/base/bittorrent/sessionimpl.cpp:
```diff
@@ -939,8 +939,8 @@ Path SessionImpl::categorySavePath(const QString &categoryName, const CategoryOp
     {
-        // use implicit save path
-        path = Utils::Fs::toValidPath(subcategoryName(categoryName));
+        QString sub = subcategoryName(categoryName);
+        path = Path(sub.isEmpty() ? QString() : Utils::Fs::toValidFileName(sub));
```
```suggestion
        const QString subcatName = subcategoryName(categoryName);
        if (!subcatName.isEmpty())
            path = Path(Utils::Fs::toValidFileName(subcatName));
```

glassez replied:
But I still have a question. How can subcategory name be empty?

cocopaw replied:
It is called recursively with the empty path being the last call.
...

glassez replied:
> It is called recursively with the empty path being the last call.

OK, I see. This phrase would be enough for me to refresh my memory of how it works.

glassez replied:
> > It is called recursively with the empty path being the last call.
> 
> OK, I see. This phrase would be enough for me to refresh my memory of how it works.

But still, subcategory name cannot be empty at this place since it returns earlier if category name is empty.

cocopaw replied:
Initially when implementing I didnt have the empty check and it caused issues.

Do you need me to recreate and show you?

glassez replied:
> Initially when implementing I didnt have the empty check and it caused issues.
> 
> Do you need me to recreate and show you?

It would be interesting to take a look at it.

cocopaw replied:
So if we remove the empty check from `categorySavePath `and just use this instead:

```
QString sub = subcategoryName(categoryName);
path = Path(Utils::Fs::toValidFileName(sub));
```

Then as soon as the path separatior "/" is typed an additional pad "_" appears after the separator.

I left the empty space check in `categoryDownloadPath `(save path for incomplete torrents section) so that you can see it working and how it should look for comparison.

<img width="490" height="198" alt="image" src="https://github.com/user-attachments/assets/5ac4ccf5-4d69-4d62-b99b-5a096d9130a4" />

<p></p>

Debug output when "boo/" is entered and the empty check has been removed. `subcategoryname `of "boo/" is empty. toValidFileName then converts empty to "_" 

```
categorySavePath() called with categoryName: "boo/"
  basePath: "C:\\Users\\HB\\Downloads\\Torrents"
  options.savePath: ""
  subcategoryName: ""
  sanitized path: "_"
categorySavePath() called with categoryName: "boo"
  basePath: "C:\\Users\\HB\\Downloads\\Torrents"
  options.savePath: ""
  subcategoryName: "boo"
  sanitized path: "boo"
categorySavePath() called with categoryName: ""
  basePath: "C:\\Users\\HB\\Downloads\\Torrents"
  categoryName is empty → returning basePath
  parent basePath: "C:\\Users\\HB\\Downloads\\Torrents"
  final result: "C:\\Users\\HB\\Downloads\\Torrents\\boo"
  parent basePath: "C:\\Users\\HB\\Downloads\\Torrents\\boo"
  final result: "C:\\Users\\HB\\Downloads\\Torrents\\boo\\_"
```

Does it make sense now?

glassez replied:
> Debug output when "boo/" is entered and the empty check has been removed. `subcategoryname `of "boo/" is empty. toValidFileName then converts empty to "_"

What the hell is this? "boo/" is an invalid category name at all!

This is what I see when I enter it into qBittorrent v5.2RC1:
<img width="499" height="209" alt="000" src="https://github.com/user-attachments/assets/1dc5050c-7646-4890-8784-14df9675ba3b" />

And that's what happens when I click OK (which is expected):
<img width="336" height="143" alt="001" src="https://github.com/user-attachments/assets/4de075bb-ab3f-41b6-a108-ff7b9197a3d0" />

cocopaw replied:
> > Debug output when "boo/" is entered and the empty check has been removed. `subcategoryname `of "boo/" is empty. toValidFileName then converts empty to "_"
> 
> What the hell is this? "boo/" is an invalid category name at all!

Nah, I'm not saying "boo/" should be a valid final category name.
It's just an intermediate state while typing something like "boo/moo".
The problem is that as soon as you type "boo/", the generated "Save path:" becomes "boo\\_" (if we remove the empty check). 
I don’t actually want to remove the empty check — I was just showing what happens without it. 
The debug output is printed on every keystroke, not when OK is pressed.

<br>

> This is what I see when I enter it into qBittorrent v5.2RC1: <img alt="000" width="499" height="209" src="https://private-user-images.githubusercontent.com/5063477/580419754-1dc5050c-7646-4890-8784-14df9675ba3b.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzY1OTQyMDEsIm5iZiI6MTc3NjU5MzkwMSwicGF0aCI6Ii81MDYzNDc3LzU4MDQxOTc1NC0xZGM1MDUwYy03NjQ2LTQ4OTAtODc4NC0xNGRmOTY3NWJhM2IucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI2MDQxOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNjA0MTlUMTAxODIxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZDNjODYzMTRjZDEwZjkyNDczN2RiNGRiOWQ2ODBiOWNlZDBjODkxZGIwZDljZTI0YjkxYzVjM2MwNjVhNzlkOSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmcmVzcG9uc2UtY29udGVudC10eXBlPWltYWdlJTJGcG5nIn0.CIjyXpe9ewrpy0sD8v3YBgRVGBJEIlV9kHmaMT-eAGE">
> 
> And that's what happens when I click OK (which is expected): <img alt="001" width="336" height="143" src="https://private-user-images.githubusercontent.com/5063477/580419760-4de075bb-ab3f-41b6-a108-ff7b9197a3d0.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzY1OTQyMDEsIm5iZiI6MTc3NjU5MzkwMSwicGF0aCI6Ii81MDYzNDc3LzU4MDQxOTc2MC00ZGUwNzViYi1hYjNmLTQxYjYtYTEwOC1mZjdiOTE5N2EzZDAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI2MDQxOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNjA0MTlUMTAxODIxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NzkyNmM5MzM2ZjFhODFiZmMxYWE3MWUyNGIxYjUwY2EzNWJmNzkxZmNjNGMwNTcwY2UwODE3Yjk2ZTQ4MmMwNCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmcmVzcG9uc2UtY29udGVudC10eXBlPWltYWdlJTJGcG5nIn0.ADWkYApBzS3OtkEKy8QTsfEDOAvKaIfZV72ZqVpiNw4">

<br>

Yes and thats also what my current implmentation on this PR with the empty check does (i.e. behaves the same way as v5.2RC1)

glassez replied:
> Nah, I'm not saying "boo/" should be a valid final category name.
> It's just an intermediate state while typing something like "boo/moo".

Well, now the context is clear. Then it definitely matters. But I would mention in the comment next to it that `categorySavePath()` can be called in the context of an incorrect category name, which is why this check is required.

glassez replied:
Also, please don't forget to apply my code change suggestion above.

cocopaw replied:
Requested changes made, Ready for further review

glassez commented on src/base/bittorrent/sessionimpl.cpp:
```diff
@@ -964,7 +964,7 @@ Path SessionImpl::categoryDownloadPath(const QString &categoryName, const Catego
     const Path path = !downloadPathOption.path.isEmpty()
             ? downloadPathOption.path
-            : Utils::Fs::toValidPath(name); // use implicit download path
+            : Path(name.isEmpty() ? QString() : Utils::Fs::toValidFileName(name)); // use implicit download path
```
Ditto (about an empty name).

cocopaw replied:
Its the same mechanism as described above.

cocopaw replied:
I changed to 'if .. else if' here to mimic your other request and avoid nested ternary operator

glassez replied:
> I changed to 'if .. else if' here to mimic your other request and avoid nested ternary operator

So why not make them exactly the same (for consistent look)?

cocopaw replied:
The OG is different and I was trying to make minimal changes, but I don't mind changing it to make it more consistent. 

Does this work better for you?

```
    if (categoryName.isEmpty())
        return downloadPath();

    Path path = downloadPathOption.path;
    if (path.isEmpty())
    {
        const QString subcatName = subcategoryName(categoryName);
        if (!subcatName.isEmpty())  // subcategoryName() returns empty string if input ends with "/"
            path = Path(Utils::Fs::toValidFileName(subcatName));
    }

    if (path.isAbsolute())
        return path;
```

glassez replied:
> Does this work better for you?

Yes. And I would preserve comment about using implicit path:
```c++
    if (categoryName.isEmpty())
        return downloadPath();

    Path path = downloadPathOption.path;
    if (path.isEmpty())
    {
        // use implicit download path
        const QString subcatName = subcategoryName(categoryName);
        if (!subcatName.isEmpty())  // subcategoryName() returns empty string if input ends with "/"
            path = Path(Utils::Fs::toValidFileName(subcatName));
    }

    if (path.isAbsolute())
        return path;
```

glassez replied:
> And I would preserve comment about using implicit path:

In both methods.

cocopaw replied:
Ok done.

Let me know if there is anything else.

glassez commented on src/base/utils/fs.cpp:
```diff
@@ -62,6 +62,89 @@
+{
+#ifdef Q_OS_WIN
+    // Set of Windows reserved device names
+    const QSet<QString> reservedDeviceNames
```
Is it used outside of `isReservedDeviceName()`?

cocopaw replied:
I moved it inside.

glassez commented on src/base/utils/fs.cpp:
```diff
@@ -62,6 +62,89 @@
+            return false;
+
+        // "." and ".." are only allowed for path components, not for filenames
+        if (isFileName && (name == u"."_s || name == u".."_s))
```
```suggestion
        if (isFileName && ((name == u"."_s) || (name == u".."_s)))
```

cocopaw replied:
done

glassez commented on src/base/utils/fs.cpp:
```diff
@@ -62,6 +62,89 @@
+    }
+
+    // Check if an individual filename or path component is valid.
+    bool isValidFsComponent(const QStringView name, bool isFileName)
```
```suggestion
    bool isValidFsComponent(const QStringView name, const bool isFileName)
```

glassez commented on src/base/utils/fs.cpp:
```diff
@@ -186,24 +269,62 @@ bool Utils::Fs::sameFiles(const Path &path1, const Path &path2)
 
-QString Utils::Fs::toValidFileName(const QString &name, const QString &pad)
+// Check if a filename is valid without sanitizing
+bool Utils::Fs::isValidFileName(QStringView name)
```
```suggestion
bool Utils::Fs::isValidFileName(const QStringView name)
```

cocopaw replied:
Done

glassez commented on src/base/utils/fs.cpp:
```diff
@@ -186,24 +269,62 @@ bool Utils::Fs::sameFiles(const Path &path1, const Path &path2)
-    return validName;
+// Check if a path component is valid without sanitizing
+// Allows "." and ".." which are valid in directory parts of a path.
+bool Utils::Fs::isValidPathComponent(QStringView component)
```
```suggestion
bool Utils::Fs::isValidPathComponent(const QStringView component)
```

glassez commented on src/base/utils/fs.cpp:
```diff
@@ -186,24 +269,62 @@ bool Utils::Fs::sameFiles(const Path &path1, const Path &path2)
+    validName.reserve(name.size());
+
+    // Replace one or more reserved characters with a single pad
+    for (const QChar c : name)
```
Maybe not so important in this case but still...
```suggestion
    for (const QChar c : asConst(name))
```

cocopaw replied:
Done

glassez commented on src/base/utils/fs.cpp:
```diff
@@ -186,24 +269,62 @@ bool Utils::Fs::sameFiles(const Path &path1, const Path &path2)
+    {
+        const qsizetype lastDotIndex = validName.lastIndexOf(u'.');
+        const QString baseName = (lastDotIndex == -1) ? validName : validName.left(lastDotIndex);
+        const QString suffix   = (lastDotIndex == -1) ? QString() : validName.sliced(lastDotIndex);
```
```suggestion
        const QString suffix = (lastDotIndex == -1) ? QString() : validName.sliced(lastDotIndex);
```

cocopaw replied:
Done

glassez commented on test/testpath.cpp:
```diff
@@ -85,11 +85,56 @@ private slots:
+        QCOMPARE(Path(u"../../../"_s).isValid(), true);
+        QCOMPARE(Path(u"./../a/../b"_s).isValid(), true);
+        QCOMPARE(Path(uR"(..\..\a\b\..)"_s).isValid(), true);
+        QCOMPARE(Path(uR"(a\.\..\b\.)"_s).isValid(), true);
```
How about testing invalid cases? E.g. `/usr/../..`?

cocopaw replied:
Ok, I've added some additional invalid test cases to cover areas related to this PR (trailing dot and space, reserved win device names etc). I also moved the control chars check from the Win to general block as these apply to all OS.  

FYI `Path("/usr/../..")` gets normalised by `QDir::cleanPath()` to `"/"` in the constructor, so `isValid()` actually returns `true`

glassez commented on src/base/utils/fs.h:
```diff
@@ -46,17 +46,17 @@ namespace Utils::Fs
 
-    bool isValidName(const QString &name);
+    bool isValidFileName(QStringView name);
+    bool isValidPathComponent(QStringView component);
```
I would consider it as redundant and just put the precondition in `Path::isValid()`:
```c++
// Split path into components and validate each one
for (const QStringView component : pathStrView.split(u'/', Qt::SkipEmptyParts))
{
    if ((component != u"."_s) && (component != u".."_s)
        && !Utils::Fs::isValidFileName(component))
    {
        return false;
    }
}
```

cocopaw replied:
Ok done.

cocopaw commented:
@glassez Thanks for the review. I can integrate those changes without any issue, but first I wanted to discuss the fundamental question:

**Should we really be allowing relative paths in isValid()?**

I agree that a general-purpose path validator should be lenient and allow relative paths, but every current call site of `Path::isValid() `is user-facing:

- webUI/API file/folder rename (`AbstractFileStorage::renameFile/renameFolder`)
- Check destination path provided is correct during torrent creation (`TorrentCreatorDialog::onCreateButtonClicked()`)
- Confirming that a usable Python installation was successfully detected (`foreignApps::PythonInfo::isValid()`)

In all of them relative paths (`.` / `..` components) are either meaningless or a source of confusion/traversal risk.

Simpler approach:

Let’s revert `isValid()` to the original stricter behaviour that disallows `.` and `..` as path components. This keeps the implementation clean and matches what all existing callers actually need.

We can always add a separate lenient helper (`isValidRelativePath() `or similar) later if a future use case requires it.

I’ll also add a clear comment on `isValid()` explaining that it is intended for user-supplied paths (not arbitrary internal paths).

Does that sound reasonable?

glassez commented:
> Does that sound reasonable?

No.

> I’ll also add a clear comment on `isValid()` explaining that it is intended for user-supplied paths (not arbitrary internal paths).

It's wrong approach.

> In all of them relative paths (`.` / `..` components) are either meaningless or a source of confusion/traversal risk.

In the cases you mentioned, it is assumed that the paths are absolute, so `.` / `..` components are unacceptable there. But they shouldn't be there, because the `Path` resolves them as much as possible, and if it's still impossible, then the path is definitely invalid (e.g. `/usr/../..`, as I mentioned above).

P.S. Of course, we could forcibly restrict all paths (including relative ones) from using `.` / `..` components, but is it worth it? Were there any problems with them doing this? I don't want to rack my brain trying to calculate the possible negative consequences, including outrage from users who may have used this for some reason.

cocopaw committed:
Glassez review changes

cocopaw committed:
Add trailing spaced check to toValidFileName

cocopaw commented:
> > Does that sound reasonable?
> 
> No.
> 
> > I’ll also add a clear comment on `isValid()` explaining that it is intended for user-supplied paths (not arbitrary internal paths).
> 
> It's wrong approach.

Ok I agree. `isValid()` should allow relative paths. 

> In the cases you mentioned, it is assumed that the paths are absolute, so `.` / `..` components are unacceptable there. But they shouldn't be there, because the `Path` resolves them as much as possible, and if it's still impossible, then the path is definitely invalid (e.g. `/usr/../..`, as I mentioned above).
> 
> P.S. Of course, we could forcibly restrict all paths (including relative ones) from using `.` / `..` components, but is it worth it? Were there any problems with them doing this? I don't want to rack my brain trying to calculate the possible negative consequences, including outrage from users who may have used this for some reason.

While relative paths (including . and .. ) are supported internally, there have been past issues where users abused GUI rename operations by entering relative directory aliases (e.g. ".", ".." or "./file.txt"). This led to unintended directory traversal via cleanPath(), subsequent folder removal, and other unintended consequences. 
That’s why I explicitly disallowed . and .. as filenames in the previous PR — to prevent such misuse. In my opinion, allowing users to input relative paths that then get resolved through `cleanPath()` may be risky, as it could theoretically enable directory traversal.
For call sites discussed on the PR like webUI/API file/folder rename, torrent creation etc, I believe we should ideally enforce absolute paths and document that expectation. However, since no actual security issues have been reported for these specific endpoints yet, I've proceeded with allowing relative paths (including . and ..) in `isValid()` for now, and leave any further hardening for a future PR if needed.

cocopaw commented:
I've implemented your suggestions and this is ready for futher review. 

As per MS guidlines, I've also added a trailing space check to `isValidFileName` and `toValidFileName`. 

[MS File/Path Naming guide](https://learn.microsoft.com/en-us/windows/win32/fileio/naming-a-file#fully-qualified-vs-relative-paths)
..
"_Do not end a file or directory name with a space or a period_"

```
    if (name.endsWith(u'.') || name.endsWith(u' '))
        return false;
```

The trailing space check together with a save path check added to `AddNewTorrentDialog::accept()` to fix issue #20412 (_trailing spaces in 'Save at' path in Add Torrent causing issues with Explorer_)

cocopaw committed:
Enhance categorySavePath\categoryDownloadPath

cocopaw committed:
Refine categorySavePath+categoryDownloadPath

glassez commented on src/base/bittorrent/sessionimpl.cpp:
```diff
@@ -939,8 +939,11 @@ Path SessionImpl::categorySavePath(const QString &categoryName, const CategoryOp
     {
-        // use implicit save path
-        path = Utils::Fs::toValidPath(subcategoryName(categoryName));
+        // use implicit download path
```
```suggestion
        // use implicit save path
```

cocopaw committed:
Update src/base/bittorrent/sessionimpl.cpp

Co-authored-by: Vladimir Golovnev <glassez@yandex.ru>

Chocobo1 commented on src/base/path.cpp:
```diff
@@ -83,34 +84,32 @@ Path::Path(const std::string &pathStr)
+    // Split path into components and validate each one
+    for (const QStringView component : pathStrView.split(u'/', Qt::SkipEmptyParts))
+    {
+        if ((component == u"."_s) || (component == u".."_s))
```
```suggestion
        if ((component == u".") || (component == u".."))
```

cocopaw replied:
Done

cocopaw replied:
Should I also remove the `_s` suffix from the `u"."` / `u".."` comparisons in `fs.cpp`?

The same pattern below appears in both `isValidFileName()` and `toValidFileName()`:

```
// The empty string, "." and ".." are not valid filenames
if (name.isEmpty() || (name == u"."_s) || (name == u".."_s))
    return pad;
```

and here too:

`validName = baseName + pad + u"1"_s + suffix;`

Chocobo1 replied:
> Should I also remove the `_s` suffix from the `u"."` / `u".."` comparisons in `fs.cpp`?
> 
> The same pattern below appears in both `isValidFileName()` and `toValidFileName()`:
> 
> ```
> // The empty string, "." and ".." are not valid filenames
> if (name.isEmpty() || (name == u"."_s) || (name == u".."_s))
>     return pad;
> ```
> 
> and here too:
> 
> `validName = baseName + pad + u"1"_s + suffix;`

Yes!

cocopaw replied:
Ok all done, tks.

Chocobo1 commented on src/base/utils/fs.h:
```diff
@@ -46,17 +46,16 @@ namespace Utils::Fs
     qint64 freeDiskSpaceOnPath(const Path &path);
 
-    bool isValidName(const QString &name);
+    bool isValidFileName(QStringView name);
```
Could you please add the forward declaration: `class QStringView;` at top of this file:
```c++
class QDateTime;
class QStringView;
```

cocopaw replied:
Done

Chocobo1 commented on src/base/utils/fs.cpp:
```diff
@@ -62,6 +62,55 @@
+        {
+            if (baseName.compare(reserved, Qt::CaseInsensitive) == 0)
+                return true;
+        }
```
```suggestion
        // add: #include <algorithm>
        return std::ranges::any_of(reservedDeviceNames, [baseName](const QString &reserved)
        {
            return baseName.compare(reserved, Qt::CaseInsensitive) == 0);
        });
```

cocopaw replied:
Done.

Chocobo1 commented on src/base/utils/fs.cpp:
```diff
@@ -186,24 +235,81 @@ bool Utils::Fs::sameFiles(const Path &path1, const Path &path2)
+
+    // Check for OS specific reserved characters
+    if (std::ranges::any_of(name, isReservedCharacter))
+        return false;
```
Can we move it to here since the preceding checks are cheaper:
```c++
#ifdef Q_OS_WIN
    // Windows restricts file names to 255 characters
    if (name.length() > 255)
        return false;

    // Windows prohibits trailing dots or spaces for filenames
    if (name.endsWith(u'.') || name.endsWith(u' '))
        return false;

    // Reserved device names (CON, PRN, AUX, etc.) cannot be used as file or directory names in Windows
    if (isReservedDeviceName(name))
        return false;
#else
    // Non-Windows systems limit file name lengths to 255 bytes in UTF-8 encoding
    if (name.toUtf8().length() > 255)
        return false;
#endif

    // Check for OS specific reserved characters
    if (std::ranges::any_of(name, isReservedCharacter))
        return false;
```

cocopaw replied:
Done

Chocobo1 reviewed (commented):
Only a few comments left.

cocopaw committed:
Chocobo1 review changes

cocopaw commented:
@Chocobo1 Thanks, changes made as requested. 

Pls let me know if there is anything else required.

Chocobo1 commented:
>Pls let me know if there is anything else required.

As soon as the [last task](https://github.com/qbittorrent/qBittorrent/pull/23386#discussion_r3142167754) is resolved and no other new issues/code change, I can approve this.

cocopaw committed:
Remove suffiix "_s"

qBittUser commented:
First commit title still has typo `enahnce` instead of `enhance`. I did mention it before: https://github.com/qbittorrent/qBittorrent/pull/23386#issuecomment-3411018133

Newest commit title also has typo, should be written `suffix`, not `suffiix`.

cocopaw commented:
> First commit title still has typo `enahnce` instead of `enhance`. I did mention it before: [#23386 (comment)](https://github.com/qbittorrent/qBittorrent/pull/23386#issuecomment-3411018133)
> 
> Newest commit title also has typo, should be written `suffix`, not `suffiix`.

Thanks.
Noted on the typos. Since this will be squash-merged the commit titles disappear anyway, so I’d rather not rewrite history for cosmetic reasons at this point.

glassez commented:
@cocopaw 
Thank you!

cocopaw commented:
@glassez @Chocobo1 
Thanks for the review!
